### ISSUE TITLE
Jest error Cannot find module 'search-expression-parser'

### ISSUE THREAD

**Issue by piassi** [MBuchalik/search-expression-parser#1]
I'm currently using this library in a typescript Next.js stack and when trying to write tests for any file that uses it, I'm getting the following error:

`Cannot find module 'search-expression-parser' from ...`

![Screen Shot 2023-04-25 at 15 39 35](https://user-images.githubusercontent.com/4440076/234371616-c2862011-65b8-430f-bc29-880f2d2e7d47.png)

Other node_modules works just fine (as you can see it is going nicely through lodash).

My jest.config.js
![Screen Shot 2023-04-25 at 15 43 26](https://user-images.githubusercontent.com/4440076/234372208-6c0c2eee-7ac3-4558-ae37-2ba8712fc86c.png)

Jest version is `28.0.3`

Any idea what the issue might be? Let me know if any additional info is needed on my side.

**Comment by MBuchalik** [comment]
My guts feeling is that your setup does not support ES Modules. Most of the (older) libraries use CommonJS modules, while `search-expression-parser` uses the newer ESM standard. So, to get started, I'd like to check whether your setup is capable of loading ES Modules.

To check this: Could you install another npm package that is also pure ESM? For instance https://github.com/sindresorhus/escape-string-regexp or https://github.com/sindresorhus/got When trying to use such a package, do you also get the same error as observed with my package?

The developer of the aforementioned packages has created a Gist with a few tips: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c

I am using the following Jest config in a project of mine (with Jest `29.5.0`), which seems to be working fine:

``` ts
import { Config } from '@jest/types';

const config: Config.InitialOptions = {
  preset: 'ts-jest/presets/default-esm',
  testEnvironment: 'node',
};

module.exports = config;
```

**Comment by piassi** [comment]
Hi @MBuchalik, thank you for taking the time to help me.

There was an ESM issue, I fixed it by following the official nextjs docs to setup my jest tests https://nextjs.org/docs/testing#jest-and-react-testing-library, but I still have the same issue with your package.

Now I have the following setup

![Screen Shot 2023-04-26 at 10 52 51](https://user-images.githubusercontent.com/4440076/234597364-a0ee9424-8400-4066-a140-3b0c6c982a4e.png)

I have wired a simple example to demonstrate that maybe the issue is not ESM related.

*With [escape-string-regexp](https://github.com/sindresorhus/escape-string-regexp):*

![Screen Shot 2023-04-26 at 10 48 26](https://user-images.githubusercontent.com/4440076/234597785-304f6c95-25d8-4e6c-852e-de78190803df.png)
Passed

*With search-expression-parser:*
![Screen Shot 2023-04-26 at 10 47 53](https://user-images.githubusercontent.com/4440076/234597611-04614ead-4e5d-4b8f-a939-116b9be59594.png)
Failed due to not finding the package

**Comment by MBuchalik** [comment]
OK I created a test installation using the following command:

```
npx create-next-app@latest --example with-jest with-jest-app
```

Then, I installed both `escape-string-regexp` and `search-expression-parser`. Both did not work due to ESM issues. So I did the following:

### Step 1
I renamed `jest.config.js` to `jest.config.mjs`

### Step 2

I updated the content of `jest.config.mjs` to the following:

``` js
import nextJest from 'next/jest.js'

const createJestConfig = nextJest({
  // Provide the path to your Next.js app to load next.config.js and .env files in your test environment
  dir: './',
})

// Add any custom config to be passed to Jest
const customJestConfig = {
  testEnvironment: 'jest-environment-jsdom',
  setupFilesAfterEnv: ['<rootDir>/jest.setup.js'],
}

const createdConfig = await createJestConfig(customJestConfig)()

// See https://github.com/vercel/next.js/issues/36077
const updatedConfig = {...createdConfig, transformIgnorePatterns: ['node_modules/(?!(escape-string-regexp|search-expression-parser|ohm-js)/)']}

export default updatedConfig
```

Please note the line at the end:

``` js
const updatedConfig = {...createdConfig, transformIgnorePatterns: ['node_modules/(?!(escape-string-regexp|search-expression-parser|ohm-js)/)']}
```

Here, I did not only add `search-expression-parser`, but also `ohm-js`. OhmJS is the parser library that is used internally, and it is also ESM based.

### Step 3

Now, I actually found a problem with my library: While I could use `escape-string-regexp`, my library would not work. So I edited `node_modules/search-expression-parser/package.json` and added the following line:

```
"exports": "./dist/index.js",
```

Now, I am able to run `search-expression-parser` in my Jest tests.

Could you test whether this approach works on your end as well? If so, then I can create a new version of `search-expression-parser` that includes the `exports` field.

If this approach does not work: Could you share a repo with me that includes a minimal setup where the issue occurs?

**Comment by piassi** [comment]
Hi @MBuchalik,

After adding the exports statement
![Screen Shot 2023-04-27 at 10 23 32](https://user-images.githubusercontent.com/4440076/234875393-6ff563b0-7938-461b-b712-db8ee496dcff.png)

 I'm getting the following error
![Screen Shot 2023-04-27 at 10 22 56](https://user-images.githubusercontent.com/4440076/234875232-a638fcd1-753f-4e94-a2f8-f037ebe9a65b.png)

**Comment by piassi** [comment]
Hi @MBuchalik,

I managed to make it work but had to remove `ohm-js` from `transformIgnorePatterns`, adding exports to the package.json worked but I ended up using `moduleNameMapper` so my code works asap.

![Screen Shot 2023-04-27 at 10 39 11](https://user-images.githubusercontent.com/4440076/234879710-49050af5-9ee0-47ff-b915-a5a1bb053fe5.png)

**Comment by MBuchalik** [comment]
Glad to hear that it works now! 🎉 I have just released version `1.0.2` which includes the `exports` field.

Interestingly enough, using `moduleNameMapper` did not work with my setup. Even in another fresh NextJS project, I had to use the approach described in https://github.com/MBuchalik/search-expression-parser/issues/1#issuecomment-1523627733. This could be caused by different versions of Jest (or other dependencies).

I will close this issue for now since it seems to be working both on my and your setup. If you experience any related problems, feel free to reopen this issue or to create a new one. 😃
